dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/maintenance/jdbc/JdbcMaintenanceStore.java

```diff
@@ -165,6 +204,31 @@ public int deleteSoftDeletedTrackedEntityInstances()
             "delete from programownershiphistory where trackedentityinstanceid in " + teiSelect,
             "delete from trackedentityinstance where deleted is true" };
 
-        return jdbcTemplate.batchUpdate( sqlStmts )[sqlStmts.length - 1];
+        int result = jdbcTemplate.batchUpdate( sqlStmts )[sqlStmts.length - 1];
+
+        if ( result > 0 && deletedTeiUids.size() > 0 )
+        {
+            auditHardDeletedTrackedEntityInstances( deletedTeiUids );
+        }
+
+        return result;
+    }
+
+    private void auditHardDeletedTrackedEntityInstances( List<String> deletedTeiUids )
+    {
+        deletedTeiUids.forEach( teiUid -> {
+            TrackedEntityInstance tei = new TrackedEntityInstance();
+            tei.setUid( teiUid );
+            tei.setDeleted( true );
+            auditManager.send( Audit.builder()
+                .auditType( AuditType.DELETE )
+                .auditScope( AuditScope.TRACKER )
+                .createdAt( LocalDateTime.now() )
+                .createdBy( usernameSupplier.get() )
```

Comment:
Design consideration: I wonder if we should allow sending an `Audit` object without the `createdBy` pre-populated (to avoid having to use the `UserSupplier` in each class that needs to send audit).
The `AuditManager` can check if the `createdAt` is null and use the supplier to populate.

The design advantage is that we remove one dependency (`UserSupplier`) from  classes that need to use to create an `Audit` object. .

The `send` method of `AuditManager` would contain this snippet:

```
if ( StringUtils.isEmpty( audit.getCreatedBy() ) )
{
    audit.setCreatedBy( usernameSupplier.get() );
}
```
@mortenoh what do you think? (this is really a minor thing)

Reply:
That makes sense for me, @vietnguyen can you have a look?

Reply:
Agree, fixed now